### DISCUSSION TITLE
Language specific syntax highlighting

### DISCUSSION THREAD

**Discussion by trvsdnn** [zed-industries/zed#13631] [Feature Requests]
I'm working on a theme, but due to the syntax keys being a little too normalized across languages I can't dial it in the way my textmate theme is/was.

Was curious if there's a way to apply syntax styles based on language or have language overrides within the theme?

**Comment by lvignoli** [comment]
You can add a `language` directory, as stated [there](https://github.com/zed-industries/extensions/blob/main/AUTHORING_EXTENSIONS.md#languages-1).

There you should put a config.toml, and use a highlight.scm file to map ranges of source codes (as AST sub-trees) to logical "highlight-names" (as dubbed [by tree-sitter](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#theme)). These "highlight names" are made available by Zed to highlight code as specified in the `syntax` key of the JSON theme definition.

Or: the syntax keys available to themes for a given language are the highlight names exported by a highlight.scm, default Zed one or provided by your extension, which itself depends on the grammar used to construct the AST.

You can blindly copy [languages folders from Zed](https://github.com/zed-industries/zed/tree/main/crates/languages/src) for your favourites one into your extension and start from there. I think that the highligh.scm shadows Zed default one, and this only when you select the corresponding theme, but you should check.

**Reply by trvsdnn** [reply]
I'd considered that as an option, but hadn't dug into it yet. Was more a question of if there's a way directly within the theme json to scope a key for a language, e.g., `"html.punctuation.bracket"`.

Feels like overkill as part of theming to jump into tree-sitter and redefine highlights. Although it does allow the theme to completely control syntax and extend it for specific languages, which is cool.

**Reply by lvignoli** [reply]
> Was more a question of if there's a way directly within the theme json to scope a key for a language, e.g., `"html.punctuation.bracket"`.

As I understand, there is no way to match to AST as simply as what you are describing. Something must link between code ranges identified through a grammar and style names used in a theme. If Zed's default highlights do not do it, it seems you must ship some along your theme.

**Reply by lvignoli** [reply]
It´s heavy for just few colors for sure, let me know if you find an alternate way!

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
